### PR TITLE
[package] add simple graph data structure

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57341 tmp intern
* #57340 [package] remove save_source_file API
* #57339 [package] remove PackageExporter.file_structure
* #57338 [package] use digraph to back dependency visualization
* **#57337 [package] add simple graph data structure**
* #57336 [package] make extern a dict
* #57335 [package] factor out `PackageExporter._get_dependencies`

Add a really simple graph data sturcutre for tracking dependencies. API
based on networkx, but I didn't want to require the dependency.

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186/)

Differential Revision: [D28114186](https://our.internmc.facebook.com/intern/diff/D28114186)

### CODE REVIEW

File: torch/package/_digraph.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    def __iter__(self):
+        """Iterate over the nodes."""
+        return iter(self._node)
+
+    def __contains__(self, n):
+        """Returns True if ``n`` is a node in the graph, False otherwise."""
+        try:
+            return n in self._node
+        except TypeError:
+            return False
```
Comment (SplitInfinity) [inline torch/package/_digraph.py:70]: If `self._node` is a dictionary, when does `n in self._node` throw `TypeError`?

**Review by SplitInfinity — COMMENTED** [review]
Consider adding docstrings, type annotations and/or tests.

File: torch/package/_digraph.py (hunk 1 above)
Comment (suo) [reply torch/package/_digraph.py:70]: when `n` is not hashable

**Comment by suo** [comment]
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114186).

**Review by SplitInfinity — APPROVED** [review]

**Pull request closed without merging**